envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyListenersFactory.kt

```diff
@@ -0,0 +1,304 @@
+                .addGrpcServices(GrpcService.newBuilder()
+                        .setEnvoyGrpc(
+                                GrpcService.EnvoyGrpc.newBuilder()
+                                        .setClusterName("envoy-control-xds")
+                        )
+                ).build()
+    }
+
+    fun createListeners(group: Group): List<Listener> {
+        if (group.listenersConfig == null) {
+            return listOf()
+        }
+        val listenersConfig = group.listenersConfig!!
+
+        val ingressListener = Listener.newBuilder()
+                .setName("ingress_listener")
+                .setAddress(
+                        Address.newBuilder().setSocketAddress(
+                                SocketAddress.newBuilder()
+                                        .setPortValue(listenersConfig.ingressPort)
+                                        .setAddress(listenersConfig.ingressHost)
+                        )
+                )
+                .addFilterChains(createIngressFilterChain(group))
+                .build()
+
+        val egressListener = Listener.newBuilder()
+                .setName("egress_listener")
+                .setAddress(
+                        Address.newBuilder().setSocketAddress(
+                                SocketAddress.newBuilder()
+                                        .setPortValue(listenersConfig.egressPort)
+                                        .setAddress(listenersConfig.egressHost)
+                        )
+                )
+                .addFilterChains(createEgressFilterChain(group))
+                .build()
+
+        return listOf(ingressListener, egressListener)
+    }
+
+    private fun createIngressFilterChain(group: Group): FilterChain {
+        return FilterChain.newBuilder()
+                .addFilters(createIngressFilter(group))
+                .build()
+    }
+
+    private fun createEgressFilterChain(group: Group): FilterChain {
+        return FilterChain.newBuilder()
+                .addFilters(createEgressFilter(group))
+                .build()
+    }
+
+    private fun createEgressFilter(group: Group): Filter {
+        val connectionManagerBuilder = HttpConnectionManager.newBuilder()
+                .setStatPrefix("egress_http")
+                .setRds(egressRds(group.ads))
+                .setHttpProtocolOptions(egressHttp1ProtocolOptions())
+
+        return createFilter(connectionManagerBuilder, egressFilters, group, "egress")
+    }
+
+    private fun createFilter(
+        connectionManagerBuilder: HttpConnectionManager.Builder,
+        filters: List<HttpFilterFactory>,
+        group: Group,
+        accessLogType: String
+    ): Filter {
+        filters.forEach {
+            val filter = it(group)
+            if (filter != null) {
+                connectionManagerBuilder.addHttpFilters(filter)
+            }
+        }
+
+        // checked in EnvoyListenersFactory#createListeners
```

Comment:
Perhaps better pass as a parameter or an intermediate struct which is already non-optional but definitely set close to the check you mention.